filesystem/filesystem.go

```diff
@@ -496,10 +502,75 @@ func (m *Mount) addMetadata(path string, md metadata.Metadata, owner *user.User)
 	return m.writeDataAtomic(path, data, owner)
 }
 
+func readFull(fd int, data []byte) (int, error) {
+	i := 0
+	for i < len(data) {
+		count, err := unix.Read(fd, data[i:])
+		if err != nil {
+			return 0, err
+		}
+		if count == 0 {
+			break
+		}
+		i += count
+	}
+	return i, nil
+}
+
+// readMetadataFileSafe gets the contents of a metadata file extra-carefully,
+// considering that it could be a malicious file created to cause a
+// denial-of-service.  Specifically, the following checks are done:
+//
+// - It must be a regular file, not another type of file like a symlink or FIFO.
+// - It must have a reasonable size (<= maxMetadataFileSize).
+//
+// Take care to avoid TOCTOU (time-of-check-time-of-use) bugs when doing these
+// tests.  Notably, we must open the file before checking the file type, as the
+// file type could change between any previous checks and the open.  When doing
+// this, O_NOFOLLOW is needed to avoid following a symlink, and O_NONBLOCK is
+// needed to avoid blocking if the file is a FIFO.  Note that this implies we
+// have to use the low-level unix.Open instead of os.OpenFile.
+//
+// This function returns the data read.
+func readMetadataFileSafe(path string) ([]byte, error) {
+	fd, err := unix.Open(path, unix.O_RDONLY|unix.O_NOFOLLOW|unix.O_NONBLOCK, 0)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open %q", path)
+	}
+	defer unix.Close(fd)
+
+	var stat unix.Stat_t
+	if err = unix.Fstat(fd, &stat); err != nil {
+		return nil, errors.Wrapf(err, "failed to stat %q", path)
+	}
+	if (stat.Mode & unix.S_IFMT) != unix.S_IFREG {
+		return nil, &ErrCorruptMetadata{path, errors.New("not a regular file")}
+	}
+	if stat.Size > maxMetadataFileSize {
+		return nil, &ErrCorruptMetadata{path, errors.New("metadata file size limit exceeded")}
+	}
+	// Clear O_NONBLOCK, since it has served its purpose when opening the
+	// file, and the behavior of reading from a regular file with O_NONBLOCK
+	// is technically unspecified.
+	if _, err = unix.FcntlInt(uintptr(fd), unix.F_SETFL, 0); err != nil {
+		return nil, errors.Wrapf(err, "failed to clear O_NONBLOCK on %q", path)
+	}
+	// Read the file contents, making sure we hit EOF at exactly stat.Size.
+	data := make([]byte, stat.Size+1)
+	count, err := readFull(fd, data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read from %q", path)
+	}
+	if int64(count) != stat.Size {
+		return nil, errors.Errorf("file %q was concurrently modified", path)
+	}
+	return data[:count], nil
```

Comment:
For normal files we could just use an [`io.LimitedReader`](https://pkg.go.dev/io#LimitedReader):
```go
r := &io.LimitedReader{R: f, N: maxMetadataFileSize}
data, err := io.ReadAll(r)
if err != nil {
  return err
}
if r.N == 0 {
  // Return a PathError about the file being too long
}
return data
```

Then we wouldn't even need to check the file size with Stat

Reply:
Done.  (But I did `N: maxMetadataFileSize + 1` to avoid an off by one bug, and I kept `ErrCorruptMetadata` for the error.)

Reply:
`io.ReadAll()` is only Go 1.16 and later.  Trying `ioutil.ReadAll()` instead.